monolake-services/src/common/route.rs

```diff
@@ -0,0 +1,388 @@
+                    .map(LoadBalancer::WeightedRandom)?
+            }
+            LoadBalanceStrategy::RoundRobin => {
+                RoundRobinSelector::new(it.map(|up| up.into_weighted_endpoint().0).collect())
+                    .map(LoadBalancer::RoundRobin)?
+            }
+            LoadBalanceStrategy::First => {
+                let Some(up) = it.next() else {
+                    return Err(LoadBalanceError::EmptyUpstream);
+                };
+                LoadBalancer::Identity(IdentitySelector(up.into_weighted_endpoint().0))
+            }
+        })
+    }
+}
+
+impl<T, A: ?Sized> Selector<A> for LoadBalancer<T> {
+    type Output<'a>
+        = &'a T
+    where
+        Self: 'a;
+    type Error = Infallible;
+
+    #[inline]
+    fn select(&self, key: &A) -> Result<Self::Output<'_>, Self::Error> {
+        match self {
+            LoadBalancer::Random(random_selector) => random_selector.select(key),
+            LoadBalancer::WeightedRandom(wr_selector) => wr_selector.select(key),
+            LoadBalancer::RoundRobin(round_robin_selector) => round_robin_selector.select(key),
+            LoadBalancer::Identity(identity_selector) => identity_selector.select(key),
+        }
+    }
+}
+
+/// Error type for SvcRoute to indicate the error from selector or service.
+#[derive(thiserror::Error, Debug)]
+pub enum SelectError<ESEL, ESVC> {
+    #[error("selector error: {0:?}")]
+    SelectorError(ESEL),
+    #[error("service error: {0:?}")]
+    ServiceError(ESVC),
+}
+
+impl<B, ESEL: HttpError<B>, ESVC: HttpError<B>> HttpError<B> for SelectError<ESEL, ESVC> {
+    #[inline]
+    fn to_response(&self) -> Option<http::Response<B>> {
+        match self {
+            SelectError::SelectorError(e) => e.to_response(),
+            SelectError::ServiceError(e) => e.to_response(),
+        }
+    }
+}
+
+/// Dispatch service based on the selector.
+///
+/// The selector's output is the target service.
+/// This is useful when you want to dispatch request to multiple pre-constructed services.
+pub struct SvcDispatch<S>(pub S);
+
+impl<SEL, R, SR, SE, SELE> Service<R> for SvcDispatch<SEL>
+where
+    SEL: Selector<R, Error = SELE>,
+    for<'a> SEL::Output<'a>: Service<R, Response = SR, Error = SE>,
+{
+    type Response = SR;
+    type Error = SelectError<SELE, SE>;
+
+    async fn call(&self, req: R) -> Result<Self::Response, Self::Error> {
+        let svc = self.0.select(&req).map_err(SelectError::SelectorError)?;
+        svc.call(req).await.map_err(SelectError::ServiceError)
+    }
+}
+
+/// Route service based on the selector.
+///
+/// Get the selector output and call the service with (Req, Out).
+pub struct SvcRoute<SEL, SVC, F> {
+    pub selector: SEL,
+    pub selector_mapper: F,
+    pub svc: SVC,
+}
+
+pub trait Mapping<In> {
+    type Out: ?Sized;
+    fn map<'a>(&self, input: &'a In) -> &'a Self::Out;
+}
+
+impl<SVC, SEL, F, R, SVCR, SVCE, CX> Service<(R, CX)> for SvcRoute<SEL, SVC, F>
```

Comment:
I think SvcDispatch and SvcRoute naming is a little confusing. Maybe we can call `SvcDispatch` as `ServiceSelector`? Since it's picking the service to execute. And `SvcRoute` to `ServiceWithSelector`?

Reply:
I renamed `SvcDispatch` to `ServiceSelector`, and `SvcRoute` to `ServiceRouter`.